### PR TITLE
Fix RequestReJIT deadlock.

### PR DESCRIPTION
**Pull request by tonyredondo**
This PR fixes the RequestReJIT deadlock by moving the call from the ModuleFinished callback to a custom not CLR thread.

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 1]
```diff
@@ -2840,15 +2841,17 @@ size_t CorProfiler::CallTarget_RequestRejitForModule(ModuleID module_id, ModuleM
...
            "]");
+
+      rejit_handler->EnqueueForRejit(module_id, methodDef);
       
       enumIterator = ++enumIterator;
     }
   }
 
   // Request the ReJIT for all integrations found in the module.
   if (!vtMethodDefs.empty()) {
```
Comment (kevingosse) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:2852]: Don't forget to remove vtMethodDefs and vtModules in the final version ;)

Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:2852]: yeah, I'm working on that, also to accept the complete array so we don't suspend the runtime for each single method.

Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:2850]: the method was changed in: a029fef

**Review by kevingosse — APPROVED** [review]
LGTM

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h [hunk 2]
```diff
@@ -119,6 +133,47 @@ class RejitHandler {
...
+
+ private:
+  static void enqueue_thread(RejitHandler* handler) {
+    auto queue = handler->rejit_queue_;
+    auto profilerInfo = handler->profilerInfo;
+
+    Info("Initializing ReJIT request thread.");
+    HRESULT hr = profilerInfo->InitializeCurrentThread();
+    if (FAILED(hr)) {
+      Warn("Call to InitializeCurrentThread fail.");
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h:156]: Is there anything different we need to do when this fails? Probably the answer is no, but I thought I would ask

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 3]
```diff
@@ -592,6 +592,7 @@ HRESULT STDMETHODCALLTYPE CorProfiler::Shutdown() {
   // to prevent it from unloading while in use
   std::lock_guard<std::mutex> guard(module_id_to_info_map_lock_);
 
+  // rejit_handler->Shutdown();
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:595]: nit: Can we delete this line?

**Review by zacharycmontoya — APPROVED** [review]
LGTM! Thank you for the fix and to @kevingosse for identifying the issue!

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h (hunk 2 above)
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h:156]: Well I don't think we can do anything in that case.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 3 above)
Comment (tonyredondo) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: I re-enabled the shutdown process in eb1e7ab

**Pull request merged**